synapse/storage/data_stores/main/registration.py

```diff
@@ -295,6 +296,56 @@ def delete_account_validity_for_user(self, user_id):
             desc="delete_account_validity_for_user",
         )
 
+    @defer.inlineCallbacks
+    def get_info_for_users(
+        self, user_ids: List[str],
+    ):
+        """Return the user info for a given set of users
+
+        Args:
+            user_ids: A list of users to return information about
+
+        Returns:
+            Deferred[Dict[str, bool]: A dictionary mapping each user ID to
+                a dict with the following keys:
+                    * expired (bool) - whether this is an expired user
+                    * deactivated (bool) - whether this is a deactivated user
+        """
+        # Get information of all our local users
+        def _get_info_for_users_txn(txn):
+            rows = []
+
+            for user_id in user_ids:
+                sql = """
+                    SELECT u.name, u.deactivated, av.expiration_ts_ms
+                    FROM users as u
+                    LEFT JOIN account_validity as av
+                    ON av.user_id = u.name
+                    WHERE u.name = ?
+                """
+
+                txn.execute(sql, (user_id,))
+                row = txn.fetchone()
+                logger.info("Got row: %s", row)
+                if row:
+                    rows.append(row)
+
+            return rows
+
+        info_rows = yield self.db.runInteraction(
+            "get_info_for_users", _get_info_for_users_txn
+        )
+
+        return {
+            user_id: {
+                "expired": (
+                    expiration is not None and self.clock.time_msec() >= expiration
```

Comment:
```suggestion
                    expiration is None or self.clock.time_msec() >= expiration
```

Reply:
Hrm, this is saying we consider a user expired if they've been deactivated?

Reply:
Oh, hmm. Not sure. Maybe @giomfo can weigh in on this.

Reply:
1. I think expiration may be None even if the account has not been deactivated (this would be very exceptional, but possible). An expiration is set automatically if there is none, but I don't remember when exactly (this was the mechanism used to define expiration on existing accounts).
2. We should not consider expired an account if expiration is undefined. If expiration is none this means the account is highlander..

In conclusion I would not change the existing code

Reply:
As what's in the code now is the same as what we had before https://github.com/matrix-org/synapse-dinsic/pull/46/files#diff-eab4747b3e826716868de430b947bfc2L169, I'm going to not take the suggestion